swarm/network/stream/snapshot_sync_test.go

```diff
@@ -0,0 +1,734 @@
+	//iterate over all nodes...
+	for c := 0; c < len(nodes); c++ {
+		//create an array of discovery node IDs
+		ids[c] = nodes[c].ID()
+		//get the kademlia overlay address from this ID
+		a := network.ToOverlayAddr(ids[c].Bytes())
+		//append it to the array of all overlay addresses
+		conf.addrs = append(conf.addrs, a)
+		//the proximity calculation is on overlay addr,
+		//the p2p/simulations check func triggers on discover.NodeID,
+		//so we need to know which overlay addr maps to which nodeID
+		conf.idToAddrMap[ids[c]] = a
+		conf.addrToIdMap[string(a)] = ids[c]
+	}
+	log.Info("Test config successfully initialized")
+
+	//only needed for healthy call when debugging
+	ppmap = network.NewPeerPot(testMinProxBinSize, ids, conf.addrs)
+
+	//define the action to be performed before the test checks: start syncing
+	action := func(ctx context.Context) error {
+		ticker := time.NewTicker(200 * time.Millisecond)
+		defer ticker.Stop()
+		for range ticker.C {
+			healthy := true
+			for _, id := range ids {
+				r := registries[id]
+				//PeerPot for this node
+				pp := ppmap[id]
+				//call Healthy RPC
+				h := r.delivery.overlay.Healthy(pp)
+				//print info
+				log.Debug(r.delivery.overlay.String())
+				log.Debug(fmt.Sprintf("IS HEALTHY: %t", h.GotNN && h.KnowNN && h.Full))
+				if !h.GotNN || !h.Full {
+					healthy = false
+					break
+				}
+			}
+			if healthy {
+				break
+			}
+		}
+
+		if history {
+			log.Info("Uploading for history")
+			//If testing only history, we upload the chunk(s) first
+			chunks, err := uploadFileToSingleNodeStore(node.ID(), chunkCount)
+			if err != nil {
+				return err
+			}
+			conf.chunks = append(conf.chunks, chunks...)
+			//finally map chunks to the closest addresses
+			mapKeysToNodes(conf)
+		}
+
+		errc := make(chan error)
+		//variables needed to wait for all subscriptions established before uploading
+		subscriptionsDone := make(chan struct{})
+
+		//now setup and start event watching in order to know when we can upload
+		ctx, watchCancel := context.WithTimeout(context.Background(), MAX_TIMEOUT*time.Second)
+		defer watchCancel()
+
+		log.Info("Setting up stream subscription")
+		// each node Subscribes to each other's swarmChunkServerStreamName
+		var wg sync.WaitGroup
+		for j, id := range ids {
+			log.Trace(fmt.Sprintf("subscribe: %d", j))
+			client, err := net.GetNode(id).Client()
+			if err != nil {
+				return err
+			}
+			wg.Add(1)
+			watchSubscriptionEvents(ctx, id, client, &wg, errc)
+
+			if log.Lvl(*loglevel) >= log.LvlDebug {
+				//uncomment this if to see only the uploader's node kademlia
+				//otherwise print all kademlias
+				//if j == idx {
+				var kt string
+				err = client.CallContext(ctx, &kt, "stream_getKad")
+				if err != nil {
+					return err
+				}
+
+				log.Debug("kad table " + node.ID().String())
+				log.Debug(kt)
+				//}
+			}
+			//watch for peers disconnecting
+			err = streamTesting.WatchDisconnections(id, client, disconnectC, quitC)
```

Comment:
this should come before the action call

Reply:
It actually can't, as it is for every client? Well, not without an additional loop for each id and to get the client again. Isn't it more efficient in this loop here?